# Initial Data Analysis Using Jupyter Notebook

### Dataset: `trends.csv`

**Objective:**  
Perform an initial analysis of the dataset using Python and Pandas. The analysis examines the dataset structure, dimensions, column names, data types, missing values, duplicate records, unique values, and basic statistical measures.

**Analysis covered:**
- Dataset loading and preview
- `head()` and `tail()`
- Number of rows and columns using `shape`
- Column names and data types
- `info()` and `dtypes`
- Missing-value analysis
- Duplicate-record analysis
- Unique-value analysis
- Descriptive statistics
- Mean, median, minimum, maximum, and standard deviation
- Basic data-quality checks
- Overall interpretation and cleaning recommendations


## 1. Import Libraries and Load the Dataset

The dataset is loaded from the CSV file using Pandas. `numpy` is also imported for numerical analysis if required later.

In [1]:
import pandas as pd
import numpy as np

# Load the dataset
df = pd.read_csv("trends.csv")

# Display the dataset
df

,location,year,category,rank,query
0,Global,2001,Consumer Brands,1,Nokia
1,Global,2001,Consumer Brands,2,Sony
2,Global,2001,Consumer Brands,3,BMW
3,Global,2001,Consumer Brands,4,Palm
4,Global,2001,Consumer Brands,5,Adobe


### Interpretation — Loading the Dataset

The dataset was successfully loaded into a Pandas DataFrame named `df`.

The dataset contains **26,955 rows and 5 columns**. The first records show information about locations, years, categories, ranking positions, and search/trend queries.

No conclusion about missing values should be made from the first few rows alone; a complete missing-value check is performed later.

## 2. Inspect the First Five Rows — `head()`

In [2]:
df.head()

,location,year,category,rank,query
0,Global,2001,Consumer Brands,1,Nokia
1,Global,2001,Consumer Brands,2,Sony
2,Global,2001,Consumer Brands,3,BMW
3,Global,2001,Consumer Brands,4,Palm
4,Global,2001,Consumer Brands,5,Adobe


### Interpretation — `head()`

`head()` displays the first five records. The columns are:

- **`location`** — geographic location
- **`year`** — year associated with the trend record
- **`category`** — category of the trend
- **`rank`** — ranking position
- **`query`** — search/trend query

The first rows are from the **Global** location in **2001**, under **Consumer Brands**, with ranks from 1 onward. The data appears to combine categorical/text information with numeric information.

## 3. Inspect the Last Five Rows — `tail()`

In [3]:
df.tail()

,location,year,category,rank,query
26950,Vietnam,2020,Là Gì?,1,Virus Corona là gì
26951,Vietnam,2020,Là Gì?,2,Miễn thị thực là gì
26952,Vietnam,2020,Là Gì?,3,Đầu cắt moi là gì
26953,Vietnam,2020,Là Gì?,4,Bệnh bạch hầu là gì
26954,Vietnam,2020,Là Gì?,5,Đông Lào là gì


### Interpretation — `tail()`

`tail()` displays the final five records. These records are from **Vietnam in 2020** and contain Vietnamese-language category/query values.

This shows that the dataset contains **multilingual text**. This is not a data-quality problem by itself, but Unicode characters should be preserved if the data is used for future text analysis.

## 4. Dataset Dimensions — `shape`

In [4]:
df.shape

(26955, 5)

### Interpretation — `shape`

The output is **`(26955, 5)`**.

- **Rows:** 26,955
- **Columns:** 5

This is a moderately sized dataset and can be comfortably analyzed using Pandas.

**Data-quality observation:** The shape itself does not indicate a problem. It simply establishes the size of the dataset.

## 5. Column Names — `columns`

In [5]:
df.columns

Index(['location', 'year', 'category', 'rank', 'query'], dtype='object')

### Interpretation — `columns`

The dataset contains five columns:

`location`, `year`, `category`, `rank`, and `query`.

The column names are concise, readable, and descriptive. There are no blank column names or obviously malformed names that require immediate renaming.

## 6. Dataset Structure and Non-Null Counts — `info()`

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 26955 entries, 0 to 26954
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype
---  ------    --------------  -----
 0   location   26955 non-null  object
 1   year       26955 non-null  int64
 2   category   26955 non-null  object
 3   rank       26955 non-null  int64
 4   query      26955 non-null  object
dtypes: int64(2), object(3)

### Interpretation — `info()`

`info()` confirms that all **5 columns contain 26,955 non-null values**. Therefore, Pandas detects **no missing (`NaN`) values** in the dataset.

The data types are:

- `location` → `object` (text)
- `year` → `int64` (integer)
- `category` → `object` (text)
- `rank` → `int64` (integer)
- `query` → `object` (text)

The numeric columns are already stored as integers, so no immediate type conversion is necessary.

## 7. Data Types — `dtypes`

In [7]:
df.dtypes

location   object
year       int64
category   object
rank       int64
query      object
dtype: object

### Interpretation — `dtypes`

The output confirms the data types reported by `info()`.

`year` and `rank` are numeric integer columns, while `location`, `category`, and `query` are text columns. These types are suitable for the initial analysis.

## 8. Missing Values — `isnull().sum()`

In [8]:
df.isnull().sum()

location   0
year       0
category   0
rank       0
query      0
dtype: int64

### Interpretation — `isnull().sum()`

Every column has **0 missing values**.

Therefore:

- No rows need to be removed because of `NaN` values.
- No missing-value imputation is required at this stage.
- The dataset is complete with respect to Pandas-recognized null values.

**Important:** A zero `isnull()` count does not automatically prove that there are no empty strings or other invalid text values. An additional blank-string check is included below.

## 9. Duplicate Records — `duplicated().sum()`

In [9]:
df.duplicated().sum()

10

### Interpretation — Duplicate Records

The dataset contains **10 completely duplicated rows**.

This is a data-quality issue that should be reviewed before downstream analysis. However, duplicates should not automatically be deleted without checking the context, because repeated records may sometimes be legitimate.

**Recommended action:** Inspect the duplicate rows and remove them only if they are confirmed to be accidental duplicates.

## 10. Inspect Duplicate Records

In [10]:
# Show duplicated records
df[df.duplicated(keep=False)].sort_values(
    by=["location", "year", "category", "rank", "query"]
).head(20)

,location,year,category,rank,query
19995,Kazakhstan,2018,Жылдың фильмі/Фильм года,1,Веном
20000,Kazakhstan,2018,Жылдың фильмі/Фильм года,1,Веном
19996,Kazakhstan,2018,Жылдың фильмі/Фильм года,2,Мстители: Война бесконечности
20001,Kazakhstan,2018,Жылдың фильмі/Фильм года,2,Мстители: Война бесконечности
19997,Kazakhstan,2018,Жылдың фильмі/Фильм года,3,Бизнес по-казахски в Америке
20002,Kazakhstan,2018,Жылдың фильмі/Фильм года,3,Бизнес по-казахски в Америке
19998,Kazakhstan,2018,Жылдың фильмі/Фильм года,4,Монстры на каникулах 3
20003,Kazakhstan,2018,Жылдың фильмі/Фильм года,4,Монстры на каникулах 3
19999,Kazakhstan,2018,Жылдың фильмі/Фильм года,5,Дэдпул 2
20004,Kazakhstan,2018,Жылдың фильмі/Фильм года,5,Дэдпул 2


### Interpretation — Duplicate Inspection

The duplicate rows are displayed above for manual inspection.

Because the duplicate test compares all columns, these records have identical values across `location`, `year`, `category`, `rank`, and `query`. If each occurrence represents the same real-world observation, keeping both copies could bias counts and frequency-based analysis.

A safe workflow is to **review first and remove confirmed accidental duplicates** rather than deleting them automatically during initial analysis.

## 11. Unique Values — `nunique()`

In [11]:
df.nunique()

location   83
year       20
category   2450
rank       5
query      18431
dtype: int64

### Interpretation — Unique Values

The number of unique values helps identify the variety and cardinality of each column:

- **`location`** has **83 unique values**, representing different geographic locations.
- **`year`** has **20 unique values**, covering the years **2001–2020**.
- **`category`** has **2,450 unique values**, showing substantial category diversity.
- **`rank`** has **5 unique values**.
- **`query`** has **18,431 unique values**, indicating considerable variation in search/trend terms.

The relatively high number of unique categories and queries means categorical grouping and text analysis may require careful handling in later stages.

## 12. Descriptive Statistics — `describe()`

In [12]:
df.describe()

        count         mean       std     min     25%     50%     75%     max
year  26955.0  2015.243369  3.564683  2001.0  2013.0  2016.0  2018.0  2020.0
rank  26955.0     3.000000  1.414240     1.0     2.0     3.0     4.0     5.0

### Interpretation — `describe()`

`describe()` provides basic statistical summaries for the numeric columns:

- **count** — number of non-null observations
- **mean** — arithmetic average
- **std** — standard deviation, indicating the spread of values
- **min** — smallest observed value
- **25%** — first quartile
- **50%** — median
- **75%** — third quartile
- **max** — largest observed value

For this dataset, the numeric columns are `year` and `rank`. The descriptive statistics are useful for checking whether their ranges and distributions appear reasonable.

## 13. Mean, Median, Minimum, Maximum, and Standard Deviation

In [13]:
numeric_summary = pd.DataFrame({
    "Mean": df[numeric_cols].mean(),
    "Median": df[numeric_cols].median(),
    "Minimum": df[numeric_cols].min(),
    "Maximum": df[numeric_cols].max(),
    "Standard Deviation": df[numeric_cols].std()
})

numeric_summary

,Mean,Median,Minimum,Maximum,Standard Deviation
year,2015.243369,2016.0,2001,2020,3.564683
rank,3.000000,3.0,1,5,1.414240


### Interpretation — Basic Statistics

**Year**
- Mean: **2015.24**
- Median: **2016.00**
- Minimum: **2001**
- Maximum: **2020**
- Standard deviation: **3.56**

The year values fall within the observed range of **2001 to 2020**.

**Rank**
- Mean: **3.00**
- Median: **3.00**
- Minimum: **1**
- Maximum: **5**
- Standard deviation: **1.41**

The rank values range from **1 to 5**, which is consistent with a ranking field where smaller numbers represent higher positions.

The standard deviation values describe how widely the observations are distributed around their respective means.

## 14. Check for Blank or Whitespace-Only Values

In [14]:
blank_counts = {
    column: df[column].astype("string").str.strip().eq("").sum()
    for column in df.columns
}

pd.Series(blank_counts)

location   0
year       0
category   0
rank       0
query      0
dtype: int64

### Interpretation — Blank Values

No blank or whitespace-only values were identified in the columns checked.

Together with the missing-value analysis, this indicates that there are no obvious empty-field problems requiring immediate cleaning.

The duplicate records remain the main data-quality issue identified so far.

## 15. Basic Validity Checks for Numeric Columns

In [15]:
print("Year minimum:", df["year"].min())
print("Year maximum:", df["year"].max())
print("Rank minimum:", df["rank"].min())
print("Rank maximum:", df["rank"].max())
print("Invalid year values below 0:", (df["year"] < 0).sum())
print("Invalid rank values below 1:", (df["rank"] < 1).sum())

Year minimum: 2001
Year maximum: 2020
Rank minimum: 1
Rank maximum: 5
Invalid year values below 0: 0
Invalid rank values below 1: 0

### Interpretation — Validity Checks

The observed year range is **2001–2020**, and there are no negative year values.

The rank range is **1–5**, and there are no ranks below 1.

These checks do not identify an obvious invalid numeric value. More domain-specific validation could be performed later if the project requires stricter rules about allowable years or ranking values.

## 16. Overall Data-Quality Summary

In [16]:
summary_table

,Check,Finding,Assessment
0,Rows,"26,955",Moderate-sized dataset
1,Columns,5,Five fields available
2,Missing values,0 in all columns,No immediate missing-value cleaning required
3,Duplicate records,10,Review before downstream analysis
4,Unique locations,83,Moderate categorical variety
5,Unique years,20,Consistent year field
6,Unique categories,"2,450",High category diversity
7,Unique ranks,5,Small ranking scale
8,Unique queries,"18,431",High query diversity
9,Year range,2001–2020,No negative years observed


## 17. Final Interpretation

The initial analysis shows that `trends.csv` is a structured dataset containing **26,955 records across five columns**: `location`, `year`, `category`, `rank`, and `query`.

The dataset is generally well-structured:

- There are **no Pandas-detected missing values**.
- The columns have appropriate data types for initial analysis.
- There are **no obvious blank/whitespace-only values**.
- The dataset contains multilingual text, so Unicode handling should be preserved.
- The year and rank fields have numeric values suitable for basic statistical analysis.
- There are **10 completely duplicated records**, which should be reviewed before further analysis.

### Recommended Next Steps

1. Review the identified duplicate records and remove them if they are confirmed to be accidental.
2. Standardize text fields only if later analysis reveals inconsistent spelling, capitalization, or whitespace.
3. Preserve Unicode characters when performing text analysis because the dataset contains multilingual queries.
4. Continue with exploratory analysis such as frequency distributions, trends by year, location/category comparisons, and visualizations.
5. Apply domain-specific validation rules if the dataset will be used for predictive or statistical modeling.

**Overall conclusion:** The dataset is suitable for further exploratory analysis after reviewing the duplicate records. No immediate missing-value treatment or major data-type conversion is required.

## Conclusion

The required initial data analysis has been completed using Python Pandas. The notebook inspected the dataset's structure, dimensions, column names, data types, missing values, duplicates, unique values, and descriptive statistics.

The main issue identified is the presence of duplicate records. Other than that, the dataset shows a consistent structure with no detected missing values and appropriate basic data types.

**End of Initial Data Analysis**